### DISCUSSION TITLE
How to parse an Intl-formatted date?

### DISCUSSION THREAD

**Discussion by ragulka** [moment/luxon#1069] [Q&A]
Given I have a custom Intl date format, how can I parse the formatted string as a date?

It seems I'd need to do something similar to the following:
```js
const dateTimeOptions = { day: '2-digit', month: '2-digit', year: 'numeric' }

const formattedDate = DateTime.now().setLocale('de').toLocaleString(dateTimeOptions);

const format = '' // how can I determine this from the dateTimeOptions?

const parsedDate = DateTime.fromFormat(formattedDate, format, { locale: 'de' }) 
```

...but I have no idea if I can map the Intl dateTimeOptions to a string with tokens?

Is there a way to achieve what I have in mind?

**Comment by icambron** [comment]
There isn't a way to do that generically in Luxon right now. We have the code for it, but it's not exposed in a useful way.

The way Luxon works is that those options are sent directly to Intl.DateTimeFormat, so we don't actually parse them into a format string as any part of the formatting process. 

However, because Luxon supports "macro" tokens in the parser (e.g. you can use `TTT` in a parsing format string to represent  `TIME_24_WITH_SHORT_OFFSET`), Luxon knows how to make this conversion. it's ugly:

 1. use a dummy date and call `formatToParts()` on it
 2. take those parts and map them individually back to tokens using a reverse-lookup table
 3. (not actually implemented anywhere in Luxon) reassemble them back into a string and return it

The code for 1 & 2 is here: https://github.com/moment/luxon/blob/master/src/impl/tokenParser.js#L383

So this is a feature request

Edit: looks like I have an in-progress PR for exposing this: https://github.com/moment/luxon/pull/840/files

**Reply by sph3rex** [reply]
Just for the sake of posterity, something like this should work ... I'm not really sure that the regex for un-quotable chars is enough and or if there are any other types. Just note that you have to manually map the types to your options:

```
const types: Record<string, string> = {
      day: 'd',
      month: 'MMM',
      year: 'yyyy',
      hour: 'hh',
      minute: 'mm',
      dayPeriod: 'a',
    }
    const parts = new Intl.DateTimeFormat(
      language,
      { year: 'numeric', month: 'short', day: 'numeric', hour: '2-digit', minute: '2-digit', hour12: true },
    ).formatToParts(0)
    const format: string[] = []
    for (let i = 0; i < parts.length; i++) {
      if (types.hasOwnProperty(parts[i].type)) {
        format.push(types[parts[i].type.toString()])
      }
      if(parts[i].type === 'literal') {
        const valueToBePushed = parts[i].value.match(/^[ :,.]+$/g) ? parts[i].value : `'${parts[i].value}'`
        format.push(valueToBePushed)
      }
    }

    return format.join('')
```
